src/NUnitFramework/framework/Internal/Execution/WorkShift.cs

```diff
@@ -161,11 +161,11 @@ private void StartWorkers()
                 worker.Idle += (s, ea) =>
                 {
                     // Quick check first using Interlocked.Decrement
-                    if (Interlocked.Decrement(ref _busyCount) == 0)
+                    if (Interlocked.Decrement(ref _busyCount) == 0 && !HasWork)
```

Comment:
Two concerns with `HasWork` outside a `lock`: 
 - In the worst case scenario, the current thread's view of `HasWork` is stale and never updates to become `false`, so the lock is never taken. We need a memory barrier like `Volatile.Read` or `Interlocked` which brings me to the next point:
 - Reading the `IList` can result in corrupted results (dirty read, tearing- I've observed this in practice) if the `IList` is being changed by another thread. Either we need an absolute guarantee that the `IList` is thread safe (which solves the first point as well), or we need to lock read as well as write access to it (which again solves the first point).